Lib/asyncore.py

```diff
@@ -191,12 +191,40 @@ def poll2(timeout=0.0, map=None):
         ready = [(fd, map[fd], flags) for fd, flags in r]
 
         for fd, obj, flags in ready:
-            if map.get(fd) is not obj:
-                # bpo-30931: obj has been closed by a previously executed
-                # handler.  Its file descriptor may have been reused by new
-                # dispatcher registered later.
-                continue
-            readwrite(obj, flags)
+            # similar to readwrite(), but check if the dispatcher was closed
+            # before calling each dispatcher
+            try:
+                if flags & select.POLLIN:
+                    if map.get(fd) is not obj:
+                        # bpo-30931: obj has been closed by a previously
+                        # executed handler.  Its file descriptor may have been
+                        # reused by new dispatcher registered later.
+                        continue
+                    obj.handle_read_event()
+
+                if flags & select.POLLOUT:
+                    if map.get(fd) is not obj:
+                        continue
+                    obj.handle_write_event()
+
+                if flags & select.POLLPRI:
+                    if map.get(fd) is not obj:
+                        continue
+                    obj.handle_expt_event()
+
+                if flags & (select.POLLHUP | select.POLLERR | select.POLLNVAL):
+                    if map.get(fd) is not obj:
+                        continue
+                    obj.handle_close()
+            except OSError as e:
+                if e.args[0] not in _DISCONNECTED:
+                    obj.handle_error()
+                else:
+                    obj.handle_close()
+            except _reraised_exceptions:
+                raise
+            except:
+                obj.handle_error()
```

Comment:
Oh sorry, I misunderstood your proposal. I was confused between handle_read_event() defined in Lib/asyncore.py and handle_read() which is user defined.

> If dispatcher was closed in handle_connect_event(), we will call handle_read() :-)

Right. Oh god, I didn't expect such many race conditions.

```
def handle_read_event(self):
    if self._fileno is None:
        raise DispatcherClosed
```

Why not just "return", do nothing if the dispatcher was closed? Existing example:

```
    def handle_write_event(self):
        if self.accepting:
            # Accepting sockets shouldn't get a write event.
            # We will pretend it didn't happen.
            return

        if not self.connected:
            if self.connecting:
                self.handle_connect_event()
        self.handle_write()
```